h2o-core/src/main/java/water/parser/FVecParseWriter.java

```diff
@@ -27,15 +26,60 @@
-    _ctypes = ctypes;           // Required not-null
-    _vecs = avs;
-    _nvs = new NewChunk[avs.length];
-    for(int i = 0; i < avs.length; ++i)
-      _nvs[i] = _vecs[i].chunkForChunkIdx(cidx);
-    _categoricals = categoricals;
-    _nCols = avs.length;
+  int[] _parse_columns_indices;
+
+  public FVecParseWriter(Vec.VectorGroup vg, int cidx, Categorical[] categoricals, byte[] ctypes, int chunkSize,
+                         AppendableVec[] avs) {
+    this(vg, cidx, categoricals, ctypes, chunkSize, avs, null);
+  }
+  // note that if parse_columns_indices==null, it implies all columns are parsed.
+  public FVecParseWriter(Vec.VectorGroup vg, int cidx, Categorical[] categoricals, byte[] ctypes, int chunkSize,
+                         AppendableVec[] avs, int[] parse_columns_indices) {
+
+    boolean ctypesShrunk = false;
+         // Required not-null
+    if ((parse_columns_indices!=null) && (categoricals!=null) &&
+            (parse_columns_indices.length == categoricals.length)) { // for nextChunk() calls in gzip/zip parser
+      _ctypes = ctypes;
+      _categoricals = categoricals;
+      _vecs = avs;
+      _parse_columns_indices = parse_columns_indices;
+      int num_parse_columns=parse_columns_indices.length;
+      _nvs = new NewChunk[num_parse_columns];
+      for (int i = 0; i < num_parse_columns; ++i) {
+        _nvs[i] = avs[i].chunkForChunkIdx(cidx);
+      }
+    } else {
+      if (parse_columns_indices == null) {
+        parse_columns_indices = new int[avs.length];
+        _ctypes = ctypes;
+        _categoricals = categoricals;
+        _vecs = avs;
+        for (int index = 0; index < avs.length; index++)
+          parse_columns_indices[index] = index;
+      } else {
+        int parseColNum = parse_columns_indices.length;
+        _ctypes = ctypes == null ? null : new byte[parseColNum]; // svmlight file can have ctypes=null
+        _categoricals = categoricals == null ? null : new Categorical[parseColNum]; // svmlight file can have categoricals==null
+        ctypesShrunk = (categoricals==null)? avs.length <= parse_columns_indices.length:
+                avs.length < categoricals.length; // may not be the same if user sets skipped_columns already.
+        _vecs = new AppendableVec[parseColNum];
+        for (int index = 0; index < parse_columns_indices.length; index++) {
+          if (ctypes != null)       // happens with SVMlight
+            _ctypes[index] = ctypes[parse_columns_indices[index]];
+          if (categoricals != null)   // happens with SVMlight
+            _categoricals[index] = categoricals[parse_columns_indices[index]]; // categoricals calculated for all columns
+          _vecs[index] = ctypesShrunk?avs[index]:avs[parse_columns_indices[index]];
+        }
+      }
+      _parse_columns_indices = parse_columns_indices;
+      int num_parse_columns = parse_columns_indices.length;
+      _nvs = new NewChunk[num_parse_columns];
```

Comment:
May use `parse_columns_indices.length` directly. Not a big deal.

Reply:
Was told from another code review that this provides small speedup?